### PR TITLE
Update visual style of CupertinoSwitch to match iOS 13

### PR DESCRIPTION
**Pull request by brandondiamond**
## Description

Updates the CupertinoSwitch widget to reflect the revised design of switches in iOS 13. In particular, the border no longer animates in response to user interaction. Instead, the background fades between colors.

**After:**
https://i.imgur.com/dMV7T5G.gif (slow 5x)
https://i.imgur.com/arKDoNk.gif (normal)

**Before:**
https://i.imgur.com/SxDYis4.gif (slow 5x)
https://i.imgur.com/k7HCNJU.gif (normal)

## Related Issues

Issue #33797

## Tests

Added a test to verify golden images of the slider in on and off states.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by LongCatIsLooong** [comment]
Hi @brandondiamond thank you for contributing, the PR looks awesome! However tests are mandatory for PRs like this. Could you add tests to ensure it works as expected? 

In case you want to add golden tests: https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package:flutter

**Comment by brandondiamond** [comment]
Thanks so much for the quick review! I'll get started on those golden tests right away. Given that this is a visual change, are there any additional tests (beyond the golden tests) that you would suggest?

**Comment by brandondiamond** [comment]
Added GIFs of before/after:

**After:**
https://i.imgur.com/dMV7T5G.gif (slow 5x)
https://i.imgur.com/arKDoNk.gif (normal)

**Before:**
https://i.imgur.com/SxDYis4.gif (slow 5x)
https://i.imgur.com/k7HCNJU.gif (normal)

**Comment by LongCatIsLooong** [comment]
@brandondiamond thanks for the GIFs!

Now that the track color change is animated, another thing we could add a test for is to verify the track color doesn't just jump between `lightBackgroundGray` and `widget.activeColor`. But in my opinion adding golden images for both active/inactive switches should be good enough.

**Review by justinmc — COMMENTED** [review]
Thanks for the PR, this is great!

+1 for tests.  I think that one golden image for on and one for off should be fine.

One very small visual nitpick.  I think that our shadow extends below the switch track when it shouldn't anymore.

|              | This Branch  | iOS 13          |
| ---------- | ----------------- | ----------------- |
| On        |  | Screen Shot 2019-07-17 at 9 52 57 AM |
| Off        |  | Screen Shot 2019-07-17 at 9 52 49 AM |

**Comment by brandondiamond** [comment]
Thanks Justin! I added a clip to ensure the thumb shadow doesn't exceed the track. I know clipping can be expensive so wanted to verify that it'd be okay in this case.

I also wanted to be mindful of antialiasing (since the clip is curved), so I used saveLayer to group the thumb drawing commands together. There's an additional save() and restore() around the clip because I wasn't certain whether there's an automatic save/restore around each paint() -- particularly for render objects that don't enforce repaint boundaries; I don't want subsequent drawing operations to be affected by the clip.

I'm very grateful for the opportunity to contribute -- thanks again for reviewing :)

**Comment by brandondiamond** [comment]
Added golden tests and generated goldens. I followed the instructions [here](https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package:flutter) but ran into permission issues when attempting to push my golden assets commit to the repo.

What's the correct way to push a change to the goldens repo without special permissions?

**Comment by justinmc** [comment]
Thanks!  I tried running this again and now that it's clipped it looks perfect to my eye.

Regarding performance and antialiasing, I'm going to track down someone that can confirm.

Regarding permissions on the goldens repo, I believe you'll have to fork that repo and open a PR like you did with this PR.  I know we tightened up this process recently so I could be wrong, but that's how I did it a few months ago.  If you post back with a link to the goldens PR, I can review it really quickly though.

**Comment by brandondiamond** [comment]
Awesome, thanks Justin :)

Here's the goldens PR:
https://github.com/flutter/goldens/pull/40

Once that goes through, I'll update this PR to reference the commit that includes the new goldens.

File: packages/flutter/test/cupertino/switch_test.dart
Comment (LongCatIsLooong) [inline packages/flutter/test/cupertino/switch_test.dart]: nit: `key` should be further indented.

File: packages/flutter/test/cupertino/switch_test.dart
Comment (LongCatIsLooong) [inline packages/flutter/test/cupertino/switch_test.dart]: nit: too much indentation.

File: packages/flutter/lib/src/cupertino/switch.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/cupertino/switch.dart]: Most of the time we just use `context.pushClipRRect`. An example:
https://github.com/flutter/flutter/blob/ab26e1b4a7cdefb5b9b2ba53e35c68ec3116f0e3/packages/flutter/lib/src/rendering/proxy_box.dart#L1371

Relaying @liyuqian 's answer, if it is just the shadow you want to clip off, it's probably not worth doing `saveLayer` (as there's no overlapping draw calls at the edge of the `RRect`, if I understand it correctly).

File: packages/flutter/test/cupertino/switch_test.dart
Comment (LongCatIsLooong) [inline packages/flutter/test/cupertino/switch_test.dart]: Could you also verify that 100 milliseconds (which is half of the full duration of the animation. Or whatever timing you prefer) into the animation, the switch still has the correct appearance? Something like this:
```
    await tester.tap(find.byKey(switchKey));
   // Kick off animation.
    await tester.pump()
    await tester.pump(const Duration(milliseconds: 100));
    await expectLater(
      find.byKey(switchKey),
      matchesGoldenFile(
        'switch.tap.turningOn.png',
        version: 0,
      ),
      skip: !isLinux,
    );
```

Comment (brandondiamond) [reply packages/flutter/test/cupertino/switch_test.dart]: This is super cool. Thanks for sharing that snippet!

File: packages/flutter/lib/src/cupertino/switch.dart
Comment (brandondiamond) [inline packages/flutter/lib/src/cupertino/switch.dart]: I wasn't sure whether to shorten this line and I'm hesitant to run dartfmt on library source (though if that's standard practice, I'll do that and re-push)

**Comment by brandondiamond** [comment]
Updated the tests, generated new goldens (ready to go in https://github.com/flutter/goldens/pull/40 -- I'll update the golden hash here once that is merged), and switched to using the PaintingContext to push a clip.

File: packages/flutter/lib/src/cupertino/switch.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/cupertino/switch.dart]: The analyzer is complaining about the missing type annotation.
```suggestion
    final Rect thumbBounds = Rect.fromLTRB(
```

File: packages/flutter/lib/src/cupertino/switch.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/cupertino/switch.dart]: The analyzer isn't happy about this line. See https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#naming-rules-for-typedefs-and-function-variables
> If you have a callback with arguments but you want to ignore the arguments, give the type and names of the arguments anyway. That way, if someone copies and pastes your code, they will not have to look up what the arguments are.


```suggestion
    context.pushClipRRect(needsCompositing, Offset.zero, thumbBounds, trackRRect, (PaintContext innerContext, Offset offset) {
```

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/cupertino/switch.dart]: Regarding dartfmt: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#formatting

> These guidelines have no technical effect, but they are still important purely for consistency and readability reasons.
> We do not yet use dartfmt. Flutter code tends to use patterns that the standard Dart formatter does not handle well. We are working with Dart team to make dartfmt aware of these patterns.

File: packages/flutter/lib/src/cupertino/switch.dart
Comment (brandondiamond) [reply packages/flutter/lib/src/cupertino/switch.dart]: Whoops, fixed!

**Comment by brandondiamond** [comment]
Thanks for the review! I'm going to read the style guide today to make sure I don't repeat these formatting mistakes [too often :)]

**Comment by brandondiamond** [comment]
Goldens merged and bumped the golden version! Should be good to go.

File: packages/flutter/lib/src/cupertino/switch.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/cupertino/switch.dart]: `PaintContext` => `PaintingContext`, sorry for the wrong suggestion.

**Comment by brandondiamond** [comment]
No worries! I've made that mistake many times

File: packages/flutter/lib/src/cupertino/switch.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/cupertino/switch.dart]:
```suggestion
    context.pushClipRRect(needsCompositing, Offset.zero, thumbBounds, trackRRect, (PaintingContext innerContext, Offset offset) {
```

File: packages/flutter/lib/src/cupertino/switch.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/cupertino/switch.dart]:
```suggestion
    context.pushClipRRect(needsCompositing, Offset.zero, thumbBounds, trackRRect, (PaintingContext innerContext, Offset offset) {
```

Comment (brandondiamond) [reply packages/flutter/lib/src/cupertino/switch.dart]: Aaaaand that's why I shouldn't quickly fix things on my phone. Sorry.

File: packages/flutter/test/cupertino/switch_test.dart
Comment (LongCatIsLooong) [inline packages/flutter/test/cupertino/switch_test.dart]: > ⏩ RUNNING: cd .; grep --line-number --extended-regexp [[:blank:]]$ packages/flutter/lib/src/cupertino/switch.dart packages/flutter/test/cupertino/switch_test.dart
packages/flutter/test/cupertino/switch_test.dart:565:    );  
🕐 ELAPSED TIME: 0:00:00.019169 for grep --line-number --extended-regexp [[:blank:]]$ 
packages/flutter/lib/src/cupertino/switch.dart packages/flutter/test/cupertino/switch_test.dart in .: 
Whitespace detected at the end of source code lines.
Please remove:

Looks like there're trailing white spaces at the end of this line. Could you remove it?

```suggestion
    );
```

**Comment by Piinks** [comment]
Currently blocked by #36690, which will resolve the golden sync issues that are causing the tests to fail here.

**Comment by brandondiamond** [comment]
Thanks for looking into this. Please let me know if there is anything I can do to help.

**Review by Piinks — COMMENTED** [review]
#36690 has landed. You may need to update your branch by pulling down the changes from master.

**Comment by brandondiamond** [comment]
@Piinks Thanks! What's the best way to do this without messing up commit history?

**Comment by Piinks** [comment]
You should be able to do a 
`git fetch upstream` 
and then 
`git pull upstream master`
This will create a commit that merges updates from master into your branch, which you can then push here. :)

**Comment by brandondiamond** [comment]
Thanks so much! Just pushed -- hopefully all looks okay :)

**Review by Piinks — APPROVED** [review]
LGTM! Thank you for the contribution! 🎉

**Review by justinmc — APPROVED** [review]
LGTM.  Thanks for this!  I'll merge soon.

**Pull request merged**
